### PR TITLE
Improve submodule performance when --use-submodules is not used

### PR DESCRIPTION
**Pull request by CosynPa**
By using the repository cache instead of directly cloning the submodules from remote.

### CODE REVIEW

File: Source/CarthageKit/Git.swift [hunk 1]
```diff
@@ -217,14 +227,24 @@ public func cloneSubmoduleInWorkingDirectory(_ submodule: Submodule, _ workingDi
...
 		}
 
-	return SignalProducer<(), CarthageError> { () -> Result<(), CarthageError> in
-		repositoryCheck("remove submodule checkout") {
-			try FileManager.default.removeItem(at: submoduleDirectoryURL)
+	let dependencyForSubmodule = Dependency.git(submodule.url)
+
+	return cloneOrFetchOnComplete(dependency: dependencyForSubmodule, preferHTTPS: preferHTTPS, commitish: submodule.sha)
+		.flatMap(.merge) { repositoryURL -> SignalProducer<String, CarthageError> in
+			return SignalProducer<(), CarthageError> { () -> Result<(), CarthageError> in
```
Comment (mdiep) [inline Source/CarthageKit/Git.swift]: It's not necessary to use `flatMap` with a custom `SignalProducer`. You can use `attemptMap` here instead.

File: Source/CarthageKit/Git.swift [hunk 2]
```diff
@@ -217,14 +227,24 @@ public func cloneSubmoduleInWorkingDirectory(_ submodule: Submodule, _ workingDi
 				}
 		}
 
-	return SignalProducer<(), CarthageError> { () -> Result<(), CarthageError> in
-		repositoryCheck("remove submodule checkout") {
-			try FileManager.default.removeItem(at: submoduleDirectoryURL)
+	let dependencyForSubmodule = Dependency.git(submodule.url)
+
+	return cloneOrFetchOnComplete(dependency: dependencyForSubmodule, preferHTTPS: preferHTTPS, commitish: submodule.sha)
```
Comment (mdiep) [inline Source/CarthageKit/Git.swift]: `Git.swift` is meant to be independent of the Carthage-specifics, so I don't think we should call `cloneOrFetch` here. If we need to do something like this here, then we should add a `referenceRepository: URL?` parameter to this function to keep it generic.

File: Source/CarthageKit/Git.swift [hunk 3]
```diff
@@ -217,14 +227,24 @@ public func cloneSubmoduleInWorkingDirectory(_ submodule: Submodule, _ workingDi
...
+					try FileManager.default.removeItem(at: submoduleDirectoryURL)
+				}
+			}
+			.then(cloneRepository(GitURL(repositoryURL.path), submoduleDirectoryURL, isBare: false))
 		}
-	}
-	.then(cloneRepository(submodule.url, workingDirectoryURL.appendingPathComponent(submodule.path), isBare: false))
-	.then(checkoutSubmodule(submodule, submoduleDirectoryURL))
-	.then(purgeGitDirectories)
+		.then(checkoutRepositoryToDirectory(submoduleDirectoryURL, submoduleDirectoryURL, force: false, revision: submodule.sha))
```
Comment (mdiep) [inline Source/CarthageKit/Git.swift]: Why is it important to have `force: false` here?

Comment (CosynPa) [reply Source/CarthageKit/Git.swift]: I'm just trying to keep the original logic. I don't know why sometimes the checkout is forced sometimes is not.

File: Source/CarthageKit/Git.swift (hunk 2 above)
Comment (CosynPa) [reply Source/CarthageKit/Git.swift]: How can it be possible when this `cloneSubmoduleInWorkingDirectory` function deals with all submodules recursively? Pass a map that maps any submodule to its cache path?

**Comment by CosynPa** [comment]
To make `Git.swift` independent of the Carthage-specifics, I've moved git related procedure of `cloneOrFetch` to `Git.swift`. The movement is in one commit, and the modification of `cloneOrFetch` in another, so reviewing individual commits should be easier than reviewing the whole PR.

I also added a `cacheURLMap` parameter to `cloneSubmoduleInWorkingDirectory`, so it is able to know how a git URL is mapped to a cache repository path without introducing any Carthage-specifics into `Git.swift`.

**Comment by tmspzz** [comment]
@CosynPa 

Do I understand correctly that you want to check if there is already a clone containing the branch & commit of the repository in Carthage's cache? If so then skip the cloning from remote?

**Comment by CosynPa** [comment]
Yes, and even if the commit does not exist in the cache repository, we only need to fetch in the cache repository, no need to clone the whole every time.

**Comment by CosynPa** [comment]
I removed the last commit which tried to solve the race condition problem because I found that commit did not solve the problem completely. The complete solution would be a little complex, so I think it's better to put it in another pull request.

**Comment by tmspzz** [comment]
@CosynPa please update the branch from master to fix the CI

**Comment by CosynPa** [comment]
Rebased onto master, but a test failed. Any idea?

**Comment by tmspzz** [comment]
Yes, it's a test trying to get the remote version. It's unrelated to you PR so for me this is good.

**Comment by tmspzz** [comment]
Thanks! 🎉

**Pull request merged**